transport/x/redis/inbound.go

```diff
@@ -0,0 +1,153 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package redis
+
+import (
+	"context"
+	"time"
+
+	"go.uber.org/yarpc/internal/errors"
+	"go.uber.org/yarpc/internal/request"
+	"go.uber.org/yarpc/serialize"
+	"go.uber.org/yarpc/transport"
+	"go.uber.org/yarpc/transport/internal"
+
+	"github.com/opentracing/opentracing-go"
+)
+
+// Inbound is a redis inbound that reads from the given queueKey. This will
+// wait for an item in the queue or until the timout is reached before trying
+// to read again.
+type Inbound struct {
+	registry transport.Registry
+	tracer   opentracing.Tracer
+
+	client        Client
+	timeout       time.Duration
+	queueKey      string
+	processingKey string
+
+	stop chan struct{}
+}
+
+// NewInbound creates a redis transport.Inbound
+func NewInbound(client Client, queueKey, processingKey string, timeout time.Duration) *Inbound {
+	return &Inbound{
+		tracer:  opentracing.GlobalTracer(),
+		client:  client,
+		timeout: timeout,
+		stop:    make(chan struct{}),
+	}
+}
+
+// WithTracer configures a tracer on this inbound.
+func (i *Inbound) WithTracer(tracer opentracing.Tracer) *Inbound {
+	i.tracer = tracer
+	return i
+}
+
+// WithRegistry configures a registry to handle incoming requests,
+// as a chained method for convenience.
+func (i *Inbound) WithRegistry(registry transport.Registry) *Inbound {
+	i.registry = registry
+	return i
+}
+
+// SetRegistry configures a registry to handle incoming requests.
+// This satisfies the transport.Inbound interface, and would be called
+// by a dispatcher when it starts.
+func (i *Inbound) SetRegistry(registry transport.Registry) {
+	i.registry = registry
+}
+
+// Start starts the inbound, reading from the queueKey
+func (i *Inbound) Start() error {
+	if i.registry == nil {
+		return errors.NoRegistryError{}
+	}
+
+	err := i.client.Start()
+	if err != nil {
+		return err
+	}
+
+	go i.start()
+	return nil
+}
+
+func (i *Inbound) start() {
+	for {
+		select {
+		case <-i.stop:
+			return
+		default:
+			i.handle()
+		}
+	}
+}
+
+// Stop ends the connection to redis
+func (i *Inbound) Stop() error {
+	close(i.stop)
+	return i.client.Stop()
```

Comment:
the client here will probably become synonymous with the transport.Transport work that @kriskowal is working on (since it's likely that a redis transport could be used for different inbounds).  Maybe make a note about how we might not need to call client.Stop here.